docker/service.go

```diff
@@ -186,15 +200,59 @@ func (s *Service) up(imageName string, create bool) error {
 	}
 
 	return s.eachContainer(func(c *Container) error {
-		if outOfSync, err := c.OutOfSync(); err != nil {
-			return err
-		} else if outOfSync {
-			logrus.Warnf("%s needs rebuilding", s.Name())
+		if create {
+			if err := s.rebuildIfNeeded(c); err != nil {
+				return err
+			}
 		}
+
 		return c.Up(imageName)
 	})
 }
 
+func (s *Service) rebuildIfNeeded(c *Container) error {
+	imageName, err := s.build()
+	if err != nil {
+		return err
+	}
+
+	if s.context.NoRecreate {
+		return nil
+	}
+
+	outOfSync, err := c.OutOfSync()
+	if err != nil {
+		return err
+	}
+
+	containerInfo, err := c.findInfo()
+	if containerInfo == nil || err != nil {
+		return err
+	}
+	name := containerInfo.Name[1:]
+
+	origRebuildLabel := containerInfo.Config.Labels[REBUILD.Str()]
+	newRebuildLabel := s.Config().Labels.MapParts()[REBUILD.Str()]
+	rebuildLabelChanged := newRebuildLabel != origRebuildLabel
+	logrus.WithFields(logrus.Fields{
+		"origRebuildLabel":    origRebuildLabel,
+		"newRebuildLabel":     newRebuildLabel,
+		"outOfSync":           outOfSync,
+		"ForceRecreate":       s.context.ForceRecreate,
+		"rebuildLabelChanged": rebuildLabelChanged}).Debug("Rebuild values")
+
+	if s.context.ForceRecreate || origRebuildLabel == "always" || rebuildLabelChanged || origRebuildLabel != "false" && outOfSync {
+		logrus.Infof("Rebuilding %s", name)
+		if _, err := c.Rebuild(imageName); err != nil {
+			return err
+		}
```

Comment:
Interesting idea. I agree it does get pretty confusing when you have different containers in a stack that have different recreate policies. I can see a scenario that is just impossible to resolve (at least with links or volumes) where a service is set to RECREATE=no and it has a dependency that is set to RECREATE=always. 

The links will be broken every time, so the RECREATE=no is basically invalid, but a lot of work to detect.

I'd feel a lot better leaving this out of libcompose (at least for now) because it can lead to these kind of impossible scenarios.  With the new parameters I think it should be possible to instrument this from outside of libcompose by setting the label in the config, and inspecting the containers before calling recreate.